peer/brontide.go

```diff
@@ -3742,3 +3761,339 @@ func (p *Brontide) attachChannelEventSubscription() error {
+	if p.isActiveChannel(chanID) {
+		p.log.Errorf("Channel(%v) is active, ignoring remove request",
+			chanID)
+		return
+	}
+
+	// The channel has not been added yet, we will log a warning as there
+	// is an unexpected call from funding manager.
+	if !p.isPendingChannel(chanID) {
+		p.log.Warnf("Channel(%v) not found, removing it anyway", chanID)
+	}
+
+	// Remove the record of this pending channel.
+	p.activeChannels.Delete(chanID)
+	p.addedChannels.Delete(chanID)
+}
+
+// sendLinkUpdateMsg sends a message that updates the channel to the
+// channel's message stream.
+func (p *Brontide) sendLinkUpdateMsg(cid lnwire.ChannelID, msg lnwire.Message) {
+	p.log.Tracef("Sending link update msg=%v", msg.MsgType())
+
+	chanStream, ok := p.activeMsgStreams[cid]
+	if !ok {
+		// If a stream hasn't yet been created, then we'll do so, add
+		// it to the map, and finally start it.
+		chanStream = newChanMsgStream(p, cid)
+		p.activeMsgStreams[cid] = chanStream
+		chanStream.Start()
+
+		// Stop the stream when quit.
+		go func() {
+			<-p.quit
+			chanStream.Stop()
+		}()
+	}
+
+	// With the stream obtained, add the message to the stream so we can
+	// continue processing message.
+	chanStream.AddMsg(msg)
+}
+
+// handleLinkUpdateMsg returns a boolean to indicate whether the given message
+// can be sent to the channel's message stream. If this message's channel is
+// still pending, it will be cached.
+func (p *Brontide) handleLinkUpdateMsg(msg lnwire.Message,
+	chanID lnwire.ChannelID) bool {
+
+	// Return true if the channel is active.
+	if p.isActiveChannel(chanID) {
+		return true
+	}
+
+	// If the channel is not active, it must be pending, otherwise this
+	// indicates an error.
+	if !p.isPendingChannel(chanID) {
+		p.log.Errorf("Received msg %s for non-existing channel %v",
+			msg.MsgType(), chanID)
+
+		return false
+	}
+
+	// Otherwise this is a pending channel, we will save the message to
+	// cache and reprocess it once the channel becomes active.
+	//
+	// Load the cache.
+	cache, err := p.earlyMsgCaches.Get(chanID)
+
+	// If only error returned from `Get` is `cache.ErrElementNotFound`,
+	// which indicates we don't have an item saved under this chanID.
+	if err != nil {
+		// Create the cache for this specific channel.
+		cache = newLnwireMsgCache(maxEarlyMsgs)
+		evicted, err := p.earlyMsgCaches.Put(chanID, cache)
+		if err != nil {
+			p.log.Errorf("Failed create cache for pending "+
+				"channel: %v, error: %v", chanID, err)
+
+			return false
+		}
+
+		p.log.Debugf("Created cache for pending channel: %v, "+
+			"evicted=%v", chanID, evicted)
+	}
+
+	// Create a cached message and save it.
+	err = cache.addMsg(&cachedlnwireMsg{
+		msg: msg,
+	})
+	if err != nil {
+		p.log.Errorf("Failed to cache msg %s for pending channel: %v, "+
+			"error: %v", msg.MsgType(), chanID, err)
+	}
+
+	p.log.Debugf("Cached early msg %s for pending channel: %v",
+		msg.MsgType(), chanID)
+
+	return false
```

Comment:
What happens if we remove all this caching code and just return true?

To me it looks like messages would automatically start queuing until the link becomes active.  If so, we could probably greatly simplify this PR.

Reply:
hmm so you mean we don't need to cache any messages here? Then what if this channel is still pending and we receive an `update_fee` from the remote?

Reply:
From what I can tell, if we return `true` from `handleLinkUpdateMsg` for pending channels, then `sendLinkUpdateMsg` will get called.  That will create a `chanStream` to queue messages until the link becomes active.

So we might not need to add all this caching stuff.

Reply:
yeah I think you are right! These messages are already saved in the queue in `chanStream` so the early cache here is unnecessary, will remove!